drltrace/drltrace.c

```diff
@@ -104,6 +105,134 @@ static app_pc exe_start;
+                    (!arg->pre && TEST(DRSYS_PARAM_OUT, arg->mode)))) {
+        ptr_uint_t deref = 0;
+        ASSERT(arg->size <= sizeof(deref), "too-big simple type");
+        /* We assume little-endian */
+        if (dr_safe_read((void *)arg->value, arg->size, &deref, NULL))
+            dr_fprintf(outf, (leading_zeroes ? " => " PFX : " => " PIFX), deref);
+    }
+}
+
+static void
+print_arg(drsys_arg_t *arg)
+{
+    dr_fprintf(outf, "\n    arg %d: ", arg->ordinal);
+    switch (arg->type) {
+    case DRSYS_TYPE_VOID:         print_simple_value(arg, true); break;
+    case DRSYS_TYPE_POINTER:      print_simple_value(arg, true); break;
+    case DRSYS_TYPE_BOOL:         print_simple_value(arg, false); break;
+    case DRSYS_TYPE_INT:          print_simple_value(arg, false); break;
+    case DRSYS_TYPE_SIGNED_INT:   print_simple_value(arg, false); break;
+    case DRSYS_TYPE_UNSIGNED_INT: print_simple_value(arg, false); break;
+    case DRSYS_TYPE_HANDLE:       print_simple_value(arg, false); break;
+    case DRSYS_TYPE_NTSTATUS:     print_simple_value(arg, false); break;
+    case DRSYS_TYPE_ATOM:         print_simple_value(arg, false); break;
+    default: {
+        if (arg->value == 0)
+            dr_fprintf(outf, "<null>");
+        else
+            dr_fprintf(outf, PFX, arg->value);
+    }
+    }
+
+    dr_fprintf(outf, " (%s%s%stype=%s%s, size=" PIFX ")",
+              (arg->arg_name == NULL) ? "" : "name=",
+              (arg->arg_name == NULL) ? "" : arg->arg_name,
+              (arg->arg_name == NULL) ? "" : ", ",
+              (arg->type_name == NULL) ? "\"\"" : arg->type_name,
+              (arg->type_name == NULL ||
+              TESTANY(DRSYS_PARAM_INLINED|DRSYS_PARAM_RETVAL, arg->mode)) ? "" : "*",
+              arg->size);
+}
+
+static bool
+drlib_iter_arg_cb(drsys_arg_t *arg, void *wrapcxt)
+{
+    if (arg->ordinal == -1)
+        return true;
+    arg->value = (ptr_uint_t)drwrap_get_arg(wrapcxt, arg->ordinal);
+
+    print_arg(arg);
+    return true; /* keep going */
+}
+
+syscall_info_t *lookup_entry(app_pc func)
+{
+    return (syscall_info_t *)hashtable_lookup(&addr2entry_table, func);
+}
+
+void print_args(app_pc func, void *wrapcxt)
+{
+    uint i;
+    void *drcontext = drwrap_get_drcontext(wrapcxt);
+    dr_fprintf(outf, "(");
+    DR_TRY_EXCEPT(drcontext, {
+        for (i = 0; i < options.all_args; i++) {
+            dr_fprintf(outf, "%s" PFX, (i != 0) ? ", " : "",
+                       drwrap_get_arg(wrapcxt, i));
+        }
+    }, {
+        dr_fprintf(outf, "<invalid memory>");
+        /* Just keep going */
+    });
+    dr_fprintf(outf, ")");
+}
+
+void print_symbolic_args(app_pc func, void *wrapcxt)
+{
+    drmf_status_t res;
+    syscall_info_t *libcall_entry;
+    libcall_entry = lookup_entry(func);
+    if (libcall_entry != NULL) {
+        /* drsys_iterate_arg_types employs static enumeration. */
+        res = drsys_iterate_arg_types((drsys_syscall_t *)libcall_entry,
```

Comment:
IMHO it's cleaner to separate things and have a lookup whose output can be used to the existing iterator, rather than having a multi-operation iterator.

BTW, if drsys_name_to_syscall() can be used statically, which should be true, then no new interface function is needed, right?  Call drsys_name_to_syscall() and pass the resulting handle to drsys_iterate_arg_types().